### PR TITLE
[WPE] Do not generate click event for long touch presses

### PR DESCRIPTION
**Pull request by Bastian-Krause**
#### 14aef58eda45d6e9a240ddff910a5234a1828d3d
<pre>
[WPE] Do not generate click event for long touch presses
<a href="https://bugs.webkit.org/show_bug.cgi?id=248076">https://bugs.webkit.org/show_bug.cgi?id=248076</a>

Reviewed by Michael Catanzaro.

Similar to the GTK port, a contextmenuevent should be generated for long
presses instead of a click event. We could generate a sequence of mouse
motion towards the desired point, mouse down, mouse up
(wpe_input_pointer_modifier_button2). This would cause the
contextmenuevent to be emitted, but also mousedown/mouseup events, which
are not expected. Since there seems to be no other way of doing this (at
the moment), implement a compromise: implement the detection of long tap
gestures, but don&apos;t generate the context mouse events yet.

Use a threshold value of 500 ms for detecting long tap gestures similar
to gtk-long-press-time&apos;s default of 500 ms.

See also: <a href="https://bugs.webkit.org/show_bug.cgi?id=251925">https://bugs.webkit.org/show_bug.cgi?id=251925</a>

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::doneWithTouchEvent): Add no-op on
TouchGestureController::ContextMenuEvent.
* Source/WebKit/UIProcess/API/libwpe/TouchGestureController.cpp:
(WebKit::TouchGestureController::handleEvent): Generate ContextMenuEvent
for long presses.
* Source/WebKit/UIProcess/API/libwpe/TouchGestureController.h: Add
ContextMenu to GesturedEvent enum, create ContextMenuEvent and add it to
EventVariant.
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::m_backend): Add no-op for
TouchGestureController::ContextMenuEvent.

Canonical link: <a href="https://commits.webkit.org/263030@main">https://commits.webkit.org/263030@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/53e364ff116e1729c45a0dce0a9fc4e71822f944

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3378 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3562 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4809 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3709 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3466 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2947 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3419 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3701 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4627 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1209 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3015 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2936 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2989 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3073 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4369 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3451 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2775 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3014 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/827 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3020 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3289 "Built successfully") | | |

### CODE REVIEW

**Comment by Bastian-Krause** [comment]
> ERROR: Source/WebKit/UIProcess/API/wpe/TouchGestureController.cpp:123:  Non-label code inside switch statements should be indented.  [whitespace/indent] [4]

I saw the style checker warning, but I don't know what to make of it. The rest of the file uses the exact same indentations for the switch statements.

**Comment by Bastian-Krause** [comment]
Gentle ping.

**Comment by Bastian-Krause** [comment]
Rebased. Ping?

File: Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp
Comment (Bastian-Krause) [inline Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:242]: Note: if #9688 is merged before this, we should pass `WebMouseEventSyntheticClickType::OneFingerTap` here to prevent emitting additional pointer events. If this is merged before, we could do that in #9688 or in a separate PR.

File: Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp
Comment (Bastian-Krause) [inline Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:250]: Same here.

**Comment by Bastian-Krause** [comment]
Looking at the [api-gtk test job](https://ews-build.webkit.org/#/builders/34/builds/82075):

> Unexpected timeouts (3)
>     [...]
>     /TestWTF
>         [...]
>         WTF_WorkQueue.DestroyDispatchedOnDispatchQueue

Is there more information available about this? Seems to be failing since my rebase. I could try to run the tests again on my side.

> Unexpected passes (4)
>     /WebKitGTK/TestUIClient
>         /webkit/WebKitWebView/mouse-target
>         /webkit/WebKitWebView/pointer-lock-permission-request
>     /WebKitGTK/TestContextMenu
>         /webkit/WebKitWebPage/context-menu
>         /webkit/WebKitWebPage/context-menu-node

I guess we can add these expected passes?

**Comment by Bastian-Krause** [comment]
After looking at the resulting events again, I guess `mousedown`/`mouseup` should not happen. But I have no idea how to prevent that without also losing the `contextmenu` event.

**Comment by Bastian-Krause** [comment]
This should probably be guarded by `#if ENABLE(CONTEXT_MENUS)`. Since the question above is not yet clear to me, I'll wait for feedback on this first.

File: Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp
Comment (Bastian-Krause) [inline Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp]: A possible compromise could be to drop the context click here completely. This way we do not create a click for long tap gestures, which should be alright for the time being. This would at least solve https://bugs.webkit.org/show_bug.cgi?id=251925 for long taps, which is my main motivation.

**Comment by Bastian-Krause** [comment]
Any feedback on this?

File: Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp
Comment (Bastian-Krause) [reply Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp]: Any feedback on this?

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp]: Your choice.

I don't know whether this change is good or not, but it seems reasonable you've had no negative feedback for five months so it's time to land it.

**Review by mcatanzaro — APPROVED** [review]

**Comment by mcatanzaro** [comment]
> Hm, I don't expect any changes for non-WPE ports, so that's strange.

This test is surely flaky.

**Comment by Bastian-Krause** [comment]
@mcatanzaro Please note the comments above before merging this.

File: Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp
Comment (Bastian-Krause) [reply Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp]: Okay, in this case I'll update this PR to drop the context click as suggested above. Thanks!

**Comment by mcatanzaro** [comment]
Yeah I noticed you need to update this. I won't merge it until you say it's ready.

File: Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp
Comment (Bastian-Krause) [reply Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp]: Done.

**Comment by Bastian-Krause** [comment]
@mcatanzaro I've updated the PR as noted above. It's ready to merge from my side. Thanks for taking the time!

File: Source/WebKit/UIProcess/API/libwpe/TouchGestureController.cpp
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/libwpe/TouchGestureController.cpp]: Oops, I just noticed this. Something's wrong here: don't make sense to assign and then immediately reassign.

Comment (Bastian-Krause) [reply Source/WebKit/UIProcess/API/libwpe/TouchGestureController.cpp]: Good catch, I'll drop the `m_gesturedEvent` assignment right before `FALLTHROUGH`.

Comment (Bastian-Krause) [reply Source/WebKit/UIProcess/API/libwpe/TouchGestureController.cpp]: Fixed. Could you have another look?

**Comment by Bastian-Krause** [comment]
@mcatanzaro Thanks for taking care of this!

**Pull request merged**
